text/0000-bench-utils.md

```diff
@@ -0,0 +1,207 @@
+context in which they are executed.
+
+The implementation of this function is backend-specific and currently requires
+inline assembly. No viable alternative is available in stable Rust.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+
+## `hint::black_box`
+
+The function:
+
+```rust
+/// An _unknown_ function that returns `x`.
+pub unsafe fn black_box<T>(x: T) -> T;
+```
+
+is an _unknown_ function, that is, a function that the compiler cannot make any
+assumptions about. It can potentially use `x` in any possible valid way that
+`unsafe` Rust code is allowed to, and requires the compiler to be maximally
+pessimistic in terms of optimizations. The compiler is still allowed to optimize
+the expression generating `x`. This function returns `x` and is a no-op in the
+virtual machine.
+
+For example ([`rust.godbolt.org`](https://godbolt.org/g/YP2GCJ)):
+
+```rust
+fn foo(x: i32) -> i32{ 
+  unsafe { hint::black_box(2 + x) };
+  3
+}
+let a = foo(2);
+```
+
+In the call to `foo(2)` the compiler is allowed to simplify the expression `2 + x` 
+down to `4`, but `4` must be stored into memory even though it is not read by
+anything afterwards because `black_box` could try to read it.
+
+### Benchmarking `Vec::push`
+
+The `hint::black_box` is useful for producing synthetic benchmarks that more
+accurately represent the behavior of a real application. In the following
+snippet, the function `bench` executes `Vec::push` 4 times in a loop:
+
+```rust
+fn push_cap(v: &mut Vec<i32>) {
+    for i in 0..4 {
+      v.push(i);
+    }
+}
+
+pub fn bench_push() -> Duration { 
+    let mut v = Vec::with_capacity(4);
+    let now = Instant::now();
+    push_cap(&mut v);
+    now.elapsed()
+}
+```
+
+Here, we allocate the `Vec`, push into it without growing its capacity, and drop
+it, without ever using it for anything. If we look at the assembly
+(https://rust.godbolt.org/z/wDckJF):
+
+
+```asm
+example::bench_push:
+  sub rsp, 24
+  call std::time::Instant::now@PLT
+  mov qword ptr [rsp + 8], rax
+  mov qword ptr [rsp + 16], rdx
+  lea rdi, [rsp + 8]
+  call std::time::Instant::elapsed@PLT
+  add rsp, 24
+  ret
+```
+
+it is pretty amazing: LLVM has actually managed to completely optimize the `Vec`
+allocation and call to `push_cap` away! In our real application, we would
+probably use the vector for something, preventing all of these optimizations
+from triggering, but in this synthetic benchmark LLVM optimizations are
+producing a benchmark that won't tell us anything about the cost of `Vec::push`.
+
+We can use `hint::black_box` to create a more realistic synthetic benchmark
+(https://rust.godbolt.org/z/CeXmxN):
+
+```rust
+fn push_cap(v: &mut Vec<i32>) {
+    for i in 0..4 {
+        unsafe {
+            black_box(v.as_ptr());
+            v.push(black_box(i));
+            black_box(v.as_ptr());
+        }
+    }
+}
+```
+
+that prevents LLVM from assuming anything about the vector across the calls to
+`Vec::push`.
```

Comment:
Actually, with `as_ptr` I am not so sure... in general, unsafe code *cannot* validly access a raw pointer. If you make this `&mut v`, I would agree.

The text should explain *why* this prevents optimization: Because `black_box` *could* do anything with the value, including observing and mutating the stack content, so LLVM has to generate code assuming the worst.

Reply:
> in general, unsafe code cannot validly access a raw pointer. 

So I mentioned this tangentially in the other comment, but IIUC the unknown function can do anything that does not invoke UB. For example, in the first iteration, the vector could be empty (if we use `Vec::new` instead of `Vec::with_capacity), so it could not dereference the raw pointer to read from it, but in the subsequent iterations, it could dereference the raw pointer to read from it because the behavior of that would be defined, and therefore, allowed.

Maybe i am coming at this with the wrong mental model, why couldn't an unsafe unknown function that is passed a valid readable pointer access it ?

Reply:
Oh, I see. Your intention is that `black_box` can do anything that unsafe code do *if it knew exactly who called it*. That's a very reasonable approach, and also answers my question further up, but it should be said more explicitly. :)

When I write a public safe function that internally uses unsafe, I couldn't do anything with that raw pointer. But when I give unsafe code full information about what my code does and then say "now you may do anything that won't break my code", *that's* much stronger, it works for this example, and it is likely what you want. So if you manage to express that in the description above, and then explain it by example down here, I am happy. :)

Reply:
@RalfJung the way I worded this is by writing:

"`black_box` is an _unknown_ function that can perform any valid operation on `x` that Rust
is allowed to perform without introducing undefined behavior in the calling
code."

I have no other ideas about how to incorporate that the calling code gives the unsafe code inside black box full information about what the calling code does in the wording :/